pkg/experiment/metastore/admin/admin.go

```diff
@@ -0,0 +1,258 @@
+		leaderCounts[node.ObservedLeader]++
+		if leaderCounts[node.ObservedLeader] > maxLeaderCount && node.ObservedLeader != "" {
+			maxLeaderCount = leaderCounts[node.ObservedLeader]
+			leaderId = node.ObservedLeader
+		}
+	}
+	return &raftNodeState{
+		Nodes:    nodes,
+		LeaderId: leaderId,
+		NumNodes: numRaftNodes,
+	}, nil
+}
+
+func (a *MetastoreAdmin) newClient(address string) (raftnodepb.RaftNodeOpsServiceClient, error) {
+	// TODO aleks-p: do we need more configuration here?
+	client, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	return raftnodepb.NewRaftNodeOpsServiceClient(client), nil
+}
+
+func (a *MetastoreAdmin) addFormActionHandlers() {
+	a.actionHandlers["add"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		_, err := a.sharedClient.AddNode(r.Context(), &raftnodepb.AddNodeRequest{ServerId: serverId})
+		return err
+	}
+	a.actionHandlers["remove"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		raftState, err := a.fetchRaftState(r.Context())
+		if err != nil {
+			return err
+		}
+		if raftState.LeaderId == serverId {
+			// In theory, this shouldn't be needed since Raft should elect a new leader upon removal.
+			// In practice, a successful leader election doesn't always occur and demoting first is safer.
+			_, err = a.sharedClient.DemoteLeader(r.Context(), &raftnodepb.DemoteLeaderRequest{ServerId: serverId})
+			if err != nil {
+				return err
+			}
+		}
+		_, err = a.sharedClient.RemoveNode(r.Context(), &raftnodepb.RemoveNodeRequest{ServerId: serverId})
+		return err
+	}
+	a.actionHandlers["promote"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		_, err := a.sharedClient.PromoteToLeader(r.Context(), &raftnodepb.PromoteToLeaderRequest{ServerId: serverId})
+		return err
+	}
+	a.actionHandlers["demote"] = func(w http.ResponseWriter, r *http.Request, serverId string) error {
+		_, err := a.sharedClient.DemoteLeader(r.Context(), &raftnodepb.DemoteLeaderRequest{ServerId: serverId})
+		return err
+	}
+}
+
+// The admin page can be open for a long time before an action is executed and the Raft state when the page is viewed
+// can be different from the current state. Therefore, any time an operation is invoked we fetch the entire state and
+// verify that the operation still makes sense.
+//
+// Alternatively, we could send along the state and verify that it hasn't changed.
+func (a *MetastoreAdmin) canPerformOperation(ctx context.Context, operation string, serverId string) error {
+	raftState, err := a.fetchRaftState(ctx)
+	if err != nil {
+		return err
+	}
+	var targetNode *metastoreNode
+	for _, node := range raftState.Nodes {
+		if node.RaftServerId == serverId {
+			targetNode = node
+			break
+		}
+	}
+	if targetNode == nil {
+		return fmt.Errorf("node %s could not be found", serverId)
+	}
+	switch operation {
```

Comment:
I should have explained myself more clearly, and I apologize. The checks in the PR are too permissive and won't be as helpful as we may want or expect. I'm not saying we don't need such validation at all (we do, but it should be implemented differently – see below). I'm saying that we should focus on a different aspect: the sanity and safety of the operations.

The thing is, a state change may happen at any given point in time: before or after the check, and before or after the operation begins. It's true that the check will work in some cases, but it's also true that it won't in others. This is a promise that can't be fulfilled and may mislead the user. I'm less concerned with outdated information the user operates on; the bigger risk arises when multiple users manage the cluster concurrently. There's a whole cluster of problems we'd need to solve to make it work (e.g., the ABA problem, resolving the happened-before relation, mutually exclusive access, etc.).

The Raft implementation we use offers some means to address this. For example, the `prevIndex` argument of the configuration change methods, which we currently ignore altogether by setting it to 0:
> If nonzero, prevIndex is the index of the only configuration upon which this change may be applied;
> if another configuration entry has been added in the meantime, this request will fail

In conjunction with the `raft leader term` check, it should address most of the issues. We could add these tokens to the form and use them like a CSRF token to ensure the control is not forged or seized in the meantime.

We do want to have _this_ check in place. I'm not sure we _need_ the permissive checks in the first place.

---

What I'm proposing is to focus on the safety of the operations: ensuring the requested operation is valid and won't harm the system. This is orthogonal to the issues outlined above: we don't care about the state transition (A -> B); we are only concerned with the target state (B). I do think we need this validation in place.

There's probably just one major case: removing the leader node. Let's handle this appropriately:
1. Transfer the leadership if the node is the leader.
2. Make the node a non-voter so it can't gain leadership while being removed.
3. Remove the node.

How we could implement this:
- It doesn't have to be done in one click or in the same handler, especially in the first version. We can reject the `RemoveNode` request if the target node is the leader. The user would have to explicitly transfer leadership (and respond to any failures that may occur).
- We can omit the second step for now. When we add it, users will be able to set a "cordon" to prevent the node from being re-elected during maintenance.
- Only when the user is confident that the cluster can operate with the remaining nodes should they remove the node.

Pseudocode of the `RemoveNode` handler check:
```go
if raft_leader_id == remove_node_request_server_id {
    return "leadership should be transferred before removing the node"
}
// We still can try to remove the leader if the remove_node_request_server_id becomes
// the leader after we obtained `raft_leader_id` and before the removal is finished.
// We should call DemoteVoter (the target node) to eliminate such situations.
```

Another problem may arise when the leader changes during demotion: we want a specific leader to step down, not the current one.

Pseudocode of the `DemoteLeader` handler check:
```go
    if this_server_id != demote_leader_request_server_id {
        return "ok"
    }
    // We never demote the wrong leader.
    // In the worst case, the demotion won't happen at all,
    // or will fail if the node loses its leadership in the meantime.
```

---

Next, we're electing the leader for displaying purposes here:
```go
 		// Each node has its own view of the world and can be out of sync with the rest.
 		// We choose the leader to be the one that is observed as the leader the most times.
 		leaderCounts[node.ObservedLeader]++
		if leaderCounts[node.ObservedLeader] > maxLeaderCount && node.ObservedLeader != "" {
 			maxLeaderCount = leaderCounts[node.ObservedLeader]
 			leaderId = node.ObservedLeader
 		}
```

I see two issues here:
1. This is an arguable way to implement leader election. It will be causing issues.
2. We don't need it as we can't rely on this information.

We want to indicate the current leader. This is not feasible due to the fact that we're trying to assemble the cluster state from requests made to different nodes at different times.

What we could be doing instead:
1. We display the information the node reports without overthinking. If nodes report different leaders, we show all them as leaders; this is the only way to catch the "split-brain" situation. `ServerId == LeaderId` is suffice. 
2. We never rely on the reported leader when performing checks or operations.

Reply:
> About VerifyLeader, I am not 100% sure about Demote. Do we want to compare the requested server id against raft.LeaderWithId() (which can also be inaccurate in my understanding) or against config.ServerId (basically, you can only demote yourself, otherwise we fail)? In either case, wouldn't VerifyLeader improve the likelihood we are making the check at the right place?

The `VerifyLeader` method might be very misleading and is very easy to misuse. When we reach the `DemoteLeader` handler, we don't have any clue on the node state and `VerifyLeader` is not helpful: the leadership may change right after the check (pretty weak check, BTW). It does not matter if the check is successful.

The only thing that matters is the node identity: the node we're transferring leadership from must be the same as specified in the request. If it is the case, we don't care much if the node is not the leader anymore – this is exactly what we want – we can even suppress errors saying that the node is not the leader.

However, we should bear in mind that there are no guarantees that the node has not been promoted the next moment after the demotion – we could call `DemoteVoter` if we'd need this guarantee (I'm curious how it's handled by raft if the node is the leader, BTW).

Reply:
I should have explained myself more clearly, and I apologize. The checks in the PR are too permissive and won't be as helpful as we may want or expect. I'm not saying we don't need such validation at all (we do, but it should be implemented differently – see below). I'm saying that we should focus on a different aspect: the sanity and safety of the operations.

The thing is, a state change may happen at any given point in time: before or after the check, and before or after the operation begins. It's true that the check will work in some cases, but it's also true that it won't in others. This is a promise that can't be fulfilled and may mislead the user. I'm less concerned with outdated information the user operates on; the bigger risk arises when multiple users manage the cluster concurrently. There's a whole cluster of problems we'd need to solve to make it work (e.g., the ABA problem, resolving the happened-before relation, mutually exclusive access, etc.).

The Raft implementation we use offers some means to address this. For example, the `prevIndex` argument of the configuration change methods, which we currently ignore altogether by setting it to 0:
> If nonzero, prevIndex is the index of the only configuration upon which this change may be applied;
> if another configuration entry has been added in the meantime, this request will fail

In conjunction with the `raft leader term` check, it should address most of the issues. We could add these tokens to the form and use them like a CSRF token to ensure the control is not forged or seized in the meantime.

We do want to have _this_ check in place. I'm not sure we _need_ these checks in the first place, in the first version.

---

What I'm proposing is to focus on the safety of the operations: ensuring the requested operation is valid and won't harm the system. This is orthogonal to the issues outlined above: we don't care about the state transition (A -> B); we are only concerned with the target state (B). I do think we need this validation in place in the first version.

There's probably just one major case: removing the leader node. Let's handle this appropriately:
1. Transfer the leadership if the node is the leader.
2. Make the node a non-voter so it can't gain leadership while being removed.
3. Remove the node.

How we could implement this:
- It doesn't have to be done in one click or in the same handler, especially in the first version. We can reject the `RemoveNode` request if the target node is the leader. The user would have to explicitly transfer leadership (and respond to any failures that may occur).
- We can omit the second step for now. When we add it, users will be able to set a "cordon" to prevent the node from being re-elected during maintenance.
- Only when the user is confident that the cluster can operate with the remaining nodes should they remove the node.

Pseudocode of the `RemoveNode` handler check:
```go
if raft_leader_id == remove_node_request_server_id {
    return "leadership should be transferred before removing the node"
}
// We still can try to remove the leader if the remove_node_request_server_id becomes
// the leader after we obtained `raft_leader_id` and before the removal is finished.
// We should call DemoteVoter (the target node) to eliminate such situations.
raft.remove(remove_node_request_server_id)
```

Another problem may arise when the leader changes during demotion: we want a specific leader to step down, not the current one.

> About VerifyLeader, I am not 100% sure about Demote. Do we want to compare the requested server id against raft.LeaderWithId() (which can also be inaccurate in my understanding) or against config.ServerId (basically, you can only demote yourself, otherwise we fail)? In either case, wouldn't VerifyLeader improve the likelihood we are making the check at the right place?

The `VerifyLeader` method might be very misleading and is very easy to misuse. When we reach the `DemoteLeader` handler, we don't have any clue on the node state and `VerifyLeader` is not helpful: the leadership may change right after the check (pretty weak check, BTW). It does not matter if the check is successful.

The only thing that matters is the node identity: the node we're transferring leadership from must be the same as specified in the request. If it is the case, we don't care much if the node is not the leader anymore – this is exactly what we want – we can even suppress errors saying that the node is not the leader.

However, we should bear in mind that there are no guarantees that the node has not been promoted the next moment after the demotion – we could call `DemoteVoter` if we'd need this guarantee (I'm curious how it's handled by raft if the node is the leader).

Pseudocode of the `DemoteLeader` handler check:
```go
if this_server_id == demote_leader_request_server_id {
     // We never demote the wrong leader.
     // In the worst case, the demotion will fail if the node
     // loses its leadership in the meantime.
     if err := raft.transfer(); err != nil {
         if !errors.Is(err, raft.ErrNotLeader) {
             return err
         }
     }
}
```

In addition, when the leader is demoted, we need to wait some time after a successful demotion. Just in case the node is going to be shut down or removed immediately after the demotion, we want our clients to discover the new leader before they start sending requests to no avail. This is a hack (see my comment about the leadership transfer), but it will protect us from many unpleasant situations.

---

Next, we're electing the leader here:
```go
// Each node has its own view of the world and can be out of sync with the rest.
// We choose the leader to be the one that is observed as the leader the most times.
leaderCounts[node.ObservedLeader]++
if leaderCounts[node.ObservedLeader] > maxLeaderCount && node.ObservedLeader != "" {
	maxLeaderCount = leaderCounts[node.ObservedLeader]
	leaderId = node.ObservedLeader
}
```

I see two issues here:
1. This is an arguable way to implement leader election. It will be causing issues.
2. We don't need it as we can't rely on this information.

We want to indicate the current leader. This is not feasible due to the fact that we're trying to assemble the cluster state from requests made to different nodes at different times.

What we could be doing instead:
1. We display the information the node reports without overthinking. If nodes report different leaders, we show all them as leaders; this is the only way to catch the "split-brain" situation. `ServerId == LeaderId` is suffice. 
2. We never rely on the reported leader when performing checks or operations.

Reply:
In an effort to reduce going back and forth, I won't respond to everything in detail. My summary from this is:
- remove the checks in the form handler (already done)
- do not allow for the current leader to be removed (it was working fine, but lets go with a simpler first iteration where you need to transfer leadership first)
- add extra checks in raft handlers, e.g., check against node identity (`config.ServerID`) where it makes sense, validate the current term and pass the config index.

I will leave the `VerifyLeader()` checks in for now. They allow us to fail early when we've lost leadership but received a request. Without this, we respond to such requests in an unintuitive way (e.g., when doing checks such as `this_server_id == demote_leader_request_server_id`).

Regarding showing who the leader is in the UI, I suggest that we keep this but phrase it as "observed leaders" with counts, so we'll clearly see when there is more than one instead of counting manually. Likewise, Promote / Demote / Remove actions are conditionally shown based on the node info that each node returned. Even if the information becomes stale, any operation will be safe to perform anyway because we have checks at the raft handler level.

Reply:
Please feel free to decline any of the suggestions. My aim is to ensure a) good user experience, and b) quality of the implementation. If you feel that both are at a good level, please proceed with merging the PR – I absolutely trust your judgment. We can always address the issues in the future.

> I will leave the VerifyLeader() checks in for now. They allow us to fail early when we've lost leadership but received a request.

I'd say it allows for "premature failure". When the user deals with cluster management, the last thing we want is to scare them with errors that they can't act on or comprehend (as these might not be failures in essence): such an experience may misdirect the user and result in incorrect decisions. The way `VerifyLeader` is used in this PR ensures our interface exhibits exactly this behavior.

We don't want to fail a request if `VerifyLeader` fails in `DemoteLeader` (and, ideally, in `PromoteLeader` if the server we're trying to promote is already the leader) – this is what was requested and the actual state is: no failure has occurred. Similarly, for adding and removing a node – we should not worry about the source and transition states: if the requested state is achieved, we should not report a failure.

The `VerifyLeader` call is not our friend in this context: its only purpose is preventing stale reads. An attempt to use it for anything else is likely a mistake. You can check how it is used in other projects (e.g., [consul's use](https://github.com/search?q=repo%3Ahashicorp%2Fconsul+VerifyLeader&type=code))

> Regarding showing who the leader is in the UI, I suggest that we keep this but phrase it as "observed leaders" with counts, so we'll clearly see when there is more than one instead of counting manually. Likewise, Promote / Demote / Remove actions are conditionally shown based on the node info that each node returned. Even if the information becomes stale, any operation will be safe to perform anyway because we have checks at the raft handler level.

👍🏻 Beware that adding the concept of "observed leader" introduces some additional cognitive load and we should address this by explaining what that means and how to use this information.

Reply:
> We don't want to fail a request if VerifyLeader fails in DemoteLeader (and, ideally, in PromoteLeader if the server we're trying to promote is already the leader) – this is what was requested and the actual state is: no failure has occurred. Similarly, for adding and removing a node – we should not worry about the source and transition states: if the requested state is achieved, we should not report a failure.

This is just my personal opinion, but if I am performing an operation on stale state - I would rather see an error, even if the state of the cluster is the desired one. The error tells me that another user made a change or the cluster itself transitioned which is relevant information - it means I tried to make an operational change on something that is no longer true. 

> Beware that adding the concept of "observed leader" introduces some additional cognitive load and we should address this by explaining what that means and how to use this information.

I think this will be clear enough, but I've added a tooltip next to the header.

Reply:
An idea for the future improvement: implement the cluster health "traffic light":
 * `green` – healthy cluster – all nodes reported the same leader and term (and config index). We need to call `ReadIndex` to make sure that the leader's term has not changed in the meantime – pretty much like a consistent read.
 * `amber` –  state transition – not all nodes reported the same information. Not `green` and not `red`.
 * `red` – degraded cluster – some of the members are not available. Later, we may add more sophisticated analysis.

What we're currently doing with observed leader and raft term is highly debatable and only barely correct.

---

> This is just my personal opinion, but if I am performing an operation on stale state – I would rather see an error, even if the state of the cluster is the desired one.

This is essentially where our perspectives diverge the most, I think.

I'm an operator that sees the interface for the first time. I'm a bit nervous because this is the first time I delve into metastore cluster management. Senior folks told me that it has some quirks and that I shouldn't worry too much if something fails (another option: "senior folks told me that it has some quirks and that I should ignore them" – this is more likely to happen or will evolve to). They said this happens from time to time and does not necessarily mean a real failure took place. But I know that the system is important, and I should exhibit extreme care.

I'm demoting the leader for maintenance. The operation fails with "request term (X) lower than raft term (Y)." I'm not entirely sure I understand what it means exactly or if I did anything wrong, but the previous leader is a follower now, and the cluster seems to operate as usual. I have a feeling that something went wrong and spend another hour trying to figure out what that was, but no luck. I think I have questions for senior engineers. Maybe I caught a bug, IDK. Maybe we should ignore such errors indeed.

This is an example of the UX our system may offer. In my opinion, we have room for improvements here. If we return an error to the user, it should be precise: does it indicate a configuration drift or a genuine operation failure?

---

> I think this will be clear enough, but I've added a tooltip next to the header.

To be entirely honest, it's not clear to me at all. I don't understand how I can use this information, but I won't insist.

Reply:
The problem you described can be solved by a translation in the UI layer from

_request term (X) lower than raft term (Y)_

to 

 _Error applying the change, the cluster state changed. Please refresh the page!_

My point that some users will want to know that their operation didn't quite work (or made sense) still stands.